src/main/xml-parsers.js

```diff
@@ -279,33 +280,83 @@ export function parseCompleteMultipart(xml) {
   }
 }
 
+const formatObjInfo = content => {
+
+  let {
+    Key,
+    LastModified,
+    ETag,
+    Size,
+    VersionId,
+    IsLatest
+  } = content
+
+  const name = toArray(Key)[0]
+  const lastModified = new Date(toArray(LastModified)[0])
+  const etag = sanitizeETag(toArray(ETag)[0])
+
+  return {
+    name,
+    lastModified,
+    etag,
+    size:Size,
+    versionId:VersionId,
+    isLatest:IsLatest
+  }
+}
+
 // parse XML response for list objects in a bucket
 export function parseListObjects(xml) {
   var result = {
     objects: [],
     isTruncated: false
   }
-  var nextMarker
+  var nextMarker, nextVersionKeyMarker
   var xmlobj = parseXml(xml)
 
-  if (!xmlobj.ListBucketResult) {
-    throw new errors.InvalidXMLError('Missing tag: "ListBucketResult"')
+  const listBucketResult = xmlobj.ListBucketResult
+  const listVersionsResult=xmlobj.ListVersionsResult
+    
+  if (listBucketResult && listBucketResult.IsTruncated) {
+    result.isTruncated = listBucketResult.IsTruncated
   }
-  xmlobj = xmlobj.ListBucketResult
-  if (xmlobj.IsTruncated) result.isTruncated = xmlobj.IsTruncated
-  if (xmlobj.Contents) {
-    toArray(xmlobj.Contents).forEach(content => {
-      var name = toArray(content.Key)[0]
-      var lastModified = new Date(toArray(content.LastModified)[0])
-      var etag = toArray(content.ETag)[0].replace(/^"/g, '').replace(/"$/g, '')
-        .replace(/^&quot;/g, '').replace(/&quot;$/g, '')
-        .replace(/^&#34;/g, '').replace(/&#34;$/g, '')
-      var size = content.Size
+  if (listVersionsResult && listVersionsResult.IsTruncated) {
+    result.isTruncated = listVersionsResult.IsTruncated
+  }
+
+  if (listBucketResult && listBucketResult.Contents) {
+    toArray(listBucketResult.Contents).forEach(content => {
+      const name = toArray(content.Key)[0]
+      const lastModified = new Date(toArray(content.LastModified)[0])
+      const etag = sanitizeETag(toArray(content.ETag)[0])
+      const size = content.Size
       result.objects.push({name, lastModified, etag, size})
-      nextMarker = name
+      if (!nextMarker) {
+        nextMarker = name
+      }
     })
   }
 
+  if (listVersionsResult && listVersionsResult.Version) {
+    toArray(listVersionsResult.Version).forEach(content => {
+      result.objects.push(formatObjInfo(content))
+    })
+  }
+
+  if (listVersionsResult && listVersionsResult.DeleteMarker) {
+    toArray(listVersionsResult.DeleteMarker).forEach(content => {
+      result.objects.push(formatObjInfo(content))
+    })
+  }
+
+  if (listVersionsResult && listVersionsResult.NextKeyMarker) {
+    nextVersionKeyMarker = listVersionsResult.NextKeyMarker
+  }
+  if (listVersionsResult && listVersionsResult.NextVersionIdMarker) {
+    result.versionIdMarker = listVersionsResult.NextVersionIdMarker
+  }
+
+
   if (xmlobj.CommonPrefixes) {
```

Comment:
It looks like you have a pretty major issue here:

You cannot use `xmlObj` here. It doesn't have `CommonPrefixes`, it has `listBucketResult` structure which has `listBucketResult.CommonPrefixes`

You have to fix this part for when `listObject` api is called without versioning info, i.e. `listQueryOpts={IncludeVersion: false}`

The same needs to be repeated for `listBucketResult.CommonPrefixes`.

Reply:
Thank you for the detailed explanation.